functionalization: fix detach()

bdhirsh opened the pull request:
`.detach()` worked in basic cases previously, but didn't properly preserve view relationships between the base and the output. This wasn't heavily tested, because autograd doesn't normally encounter `FunctionalTensorWrapper` directly, but could become more common if we fuse functionalization and autograd into a single tracing pass.

This will also be a bug fix for LTC (and XLA when they use functionalization)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87610
* #87751
* __->__ #87750

bdhirsh committed:
functionalization: fix detach()

[ghstack-poisoned]

ezyang commented:
Looking over the FunctionalTensorWrapper constructor, it doesn't seem to propagate level or generation.

ezyang commented:
You can get a test from https://github.com/pytorch/pytorch/pull/84246/commits/b6d8182158db0f1b43ffbed74571f7c5ac65ca95

bdhirsh commented:
I think you're right r.e. your earlier comment on "which constructor to use".

I forgot that `copy_tensor_metadata()` is responsible for directly copying the storage. So I'll update this PR to use the non-view constructor for `FunctionalTensorWrapper`, and manually copy the other fields instead (generation, level, view_metas_).

bdhirsh committed:
Update on "functionalization: fix detach()"

`.detach()` worked in basic cases previously, but didn't properly preserve view relationships between the base and the output. This wasn't heavily tested, because autograd doesn't normally encounter `FunctionalTensorWrapper` directly, but could become more common if we fuse functionalization and autograd into a single tracing pass.

This will also be a bug fix for LTC (and XLA when they use functionalization)




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix detach()"

`.detach()` worked in basic cases previously, but didn't properly preserve view relationships between the base and the output. This wasn't heavily tested, because autograd doesn't normally encounter `FunctionalTensorWrapper` directly, but could become more common if we fuse functionalization and autograd into a single tracing pass.

This will also be a bug fix for LTC (and XLA when they use functionalization)




[ghstack-poisoned]

bdhirsh committed:
Update on "functionalization: fix detach()"

`.detach()` worked in basic cases previously, but didn't properly preserve view relationships between the base and the output. This wasn't heavily tested, because autograd doesn't normally encounter `FunctionalTensorWrapper` directly, but could become more common if we fuse functionalization and autograd into a single tracing pass.

This will also be a bug fix for LTC (and XLA when they use functionalization)




[ghstack-poisoned]

anjali411 commented on test/test_functionalization.py:
```diff
@@ -101,6 +102,56 @@ def f(x):
         self.assert_functionalization(f, torch.ones(4))
 
+    def test_view_clone_view_inplace(self):
+        def f(input):
```
could you please add a comment for what we are testing here? (what was the bug and how does the fix resolve it ...)
